jina/types/document/graph.py

```diff
@@ -0,0 +1,339 @@
+from typing import Optional, Iterator, Tuple, Dict, Iterable
+
+import numpy as np
+
+from . import Document, DocumentSourceType
+from ..arrays import ChunkArray
+from ..struct import StructView
+from ..ndarray.sparse.scipy import SparseNdArray
+from ...importer import ImportExtensions
+from ...logging.predefined import default_logger
+
+__all__ = ["GraphDocument"]
+
+if False:
+    from scipy.sparse import coo_matrix
+
+
+class GraphDocument(Document):
+    """
+    :class:`GraphDocument` is a data type created based on Jina primitive data type :class:`Document`.
+
+    It adds functionality that lets you work with a `Document` as a `graph` where all its chunks are the nodes in the `graph`.
+
+    It exposes functionality to access and manipulate `graph related info` from the `DocumentProto` such as adjacency and edge features.
+
+    .. warning::
+        - It assumes that every ``chunk`` of a ``document`` is a node of a graph.
+
+    :param document: the document to construct from. If ``bytes`` is given
+            then deserialize a :class:`DocumentProto`; ``dict`` is given then
+            parse a :class:`DocumentProto` from it; ``str`` is given, then consider
+            it as a JSON string and parse a :class:`DocumentProto` from it; finally,
+            one can also give `DocumentProto` directly, then depending on the ``copy``,
+            it builds a view or a copy from it.
+    :param copy: when ``document`` is given as a :class:`DocumentProto` object, build a
+            view (i.e. weak reference) from it or a deep copy from it.
+    :param kwargs: further key value arguments
+    """
+
+    def __init__(
+        self,
+        document: Optional[DocumentSourceType] = None,
+        copy: bool = False,
+        **kwargs,
+    ):
+        GraphDocument._check_installed_array_packages()
+        super().__init__(document=document, copy=copy, **kwargs)
+        self._node_id_to_offset = {
+            node.id: offset for offset, node in enumerate(self.nodes)
+        }
+
+    @staticmethod
+    def _check_installed_array_packages():
+        from ... import JINA_GLOBAL
+
+        if JINA_GLOBAL.scipy_installed is None:
+            JINA_GLOBAL.scipy_installed = False
+            with ImportExtensions(
+                required=False,
+                pkg_name="scipy",
+                help_text=f"GraphDocument needs scipy ",
+            ):
+                import scipy
+
+                JINA_GLOBAL.scipy_installed = True
+
+    def add_node(self, node: "Document"):
+        """
+        Add a a node to the graph
+
+        :param node: the node to be added to the graph
+        """
+        if node.id in self._node_id_to_offset:
+            default_logger.warning(f"Document {node.id} is already a node of the graph")
+            return
```

Comment:
not need. to return anything

Reply:
either this or put an `else`, but we do not want to add `Document` again